### ISSUE TITLE
LEN, ODN, and Joint training result looks odd

### ISSUE THREAD

**Issue by pidan1231239** [yinyunie/Total3DUnderstanding#19]
Hi Yinyu:

I tried LEN, ODN training code with batch size 32, but get the following loss curve:
![image](https://user-images.githubusercontent.com/9084912/95763550-bcd23600-0ce1-11eb-815d-debc05f5ad15.png)
It looks like the test loss stops converging soon after a few epochs.
All the test loss shows below:
![image](https://user-images.githubusercontent.com/9084912/95763707-fc991d80-0ce1-11eb-8130-825551f51e39.png)

I then tried joint training with the best LEN and ODN I have and the downloaded pretrained MGNet (Total3D_downloaded_mgnet):
![image](https://user-images.githubusercontent.com/9084912/95763669-ebe8a780-0ce1-11eb-8c90-399f443541a0.png)
The test results of IoU compared to Total3D without joint training (Total3D_mgnet_beforejoint) and the downloaded pretrained Total3D (Total3D downloaded):
![image](https://user-images.githubusercontent.com/9084912/95800573-4574c480-0d2a-11eb-87b9-bb072297e69f.png)

The results shows that joint training with my best LEN and ODN makes 3D Layout IoU worsen. However it's slightly better than the downloaded Total3D which is close to the results of the paper.

Also, It seems that only IoU is provided in the testing results. How can I get the mean absolute error of cam pose, mAP of 3D detection, object translation, rotation and scale errors same as the paper?
![image](https://user-images.githubusercontent.com/9084912/95801131-f9c31a80-0d2b-11eb-8711-c8c730da1904.png)
![image](https://user-images.githubusercontent.com/9084912/95801140-fc257480-0d2b-11eb-906f-6938a1c1fcaf.png)
![image](https://user-images.githubusercontent.com/9084912/95801143-fdef3800-0d2b-11eb-9597-7a7c50394669.png)

**Comment by yinyunie** [comment]
Hi,

For the ODN and LEN training,  SUN-RGBD is of a small-scale dataset (~5000 images for training), it indeed overfits with training too many epochs, but we did not see such a serious overfitting problem. We adopt similar learning targets and evaluation metrics as https://github.com/thusiyuan/cooperative_scene_parsing. There would not be too much difference in the test curves, I think you could also have a check.

**Comment by pidan1231239** [comment]
I'll check it out. Thanks a lot!

**Comment by pidan1231239** [comment]
I tried the layout_evaluate code from cooperative_scene_parsing on the test results of the downloaded checkpoint with essential changes to path settings. However, I got 0 IoU from every sample. Are there any details that I missed? Is it possible to opensource the full evaluation code of the paper?

**Comment by Wi-sc** [comment]
Hi,

I also have some problems about the training. I tried to use downloaded pretrained models and my pretrained models to initialize the framework before joint training, but the results decrease both. Can you help me check the config file?

Downloaded pretrained models to initialize joint training:
layout_iou: 0.440098
iou_3d: 0.109684
iou_2d: 0.594273
pitch_mae: 6.258617
roll_mae: 6.970254

My pretrained models to initialize joint training:
layout_iou: 0.456693
iou_3d: 0.118330
iou_2d: 0.599362

This is my config setting:

method: TOTAL3D
resume: False
finetune: True
weight: ['out/pretrained_models/pretrained_model.pth', 'out/pretrained_models/meshnet_model.pth']
seed: 123
device:
  use_gpu: True
  num_workers: 2
data:
  dataset: sunrgbd
  split: data/sunrgbd/splits
  tmn_subnetworks: 2
  face_samples: 1
  with_edge_classifier: True
model:
  layout_estimation:
    method: PoseNet
    loss: PoseLoss
  object_detection:
    method: Bdb3DNet
    loss: DetLoss
  mesh_reconstruction:
    method: DensTMNet
    loss: ReconLoss
optimizer:
  method: Adam
  lr: 1e-4
  betas: [0.9, 0.999]
  eps: 1e-08
  weight_decay: 1e-04
scheduler:
  patience: 5
  factor: 0.5
  threshold: 0.01
train:
  epochs: 200
  phase: 'joint' # 'layout_estimation' or 'object_detection' or 'joint'.
  freeze: ['mesh_reconstruction']
  batch_size: 3
test:
  phase: 'joint'
  batch_size: 3
demo:
  phase: 'joint'
log:
  vis_path: visualization
  save_results: True
  vis_step: 100
  print_step: 50
  path: out/total3d

**Comment by longbowzhang** [comment]
Hi @chengzhag,

Sorry to bother you. I am also trying to reproduce the results presented in the paper.
Now I am working on the first stage, which is training LEN and ODN individually. But I fail to reproduce the results. For example, '3D Layout' in table 1 says 57.6 w/o joint training. But the highest value I can get is 48.9. Therefore, I am wondering whether you could kindly share your config. file?

Best,
Longbow
